src/Lean/Server/Completion.lean

```diff
@@ -81,39 +81,53 @@ open Elab
 open Meta
 open FuzzyMatching
 
+abbrev EligibleHeaderDecls := HashMap Name ConstantInfo
+
 /-- Cached header declarations for which `allowCompletion headerEnv decl` is true. -/
-builtin_initialize eligibleHeaderDeclsRef : IO.Ref (HashMap Name ConstantInfo) ← IO.mkRef {}
+builtin_initialize eligibleHeaderDeclsRef : IO.Ref (Option EligibleHeaderDecls) ←
+  IO.mkRef none
 
-/-- Caches the declarations in the header for which `allowCompletion headerEnv decl` is true. -/
-def fillEligibleHeaderDecls (headerEnv : Environment) : IO Unit := do
-  eligibleHeaderDeclsRef.modify fun startEligibleHeaderDecls =>
-    (·.2) <| StateT.run (m := Id) (s := startEligibleHeaderDecls) do
-      headerEnv.constants.forM fun declName c => do
+/--
+Returns the declarations in the header for which `allowCompletion env decl` is true, caching them
+if not already cached.
+-/
+def getEligibleHeaderDecls (env : Environment) : IO EligibleHeaderDecls := do
+  if let some eligibleHeaderDecls ← eligibleHeaderDeclsRef.get then
+    return eligibleHeaderDecls
+
+  -- enter critical section to avoid running the computation multiple times
+  -- `take` safety: we call `set` in all branches, without a chance for exceptions in between
+  if let some eligibleHeaderDecls ← unsafe eligibleHeaderDeclsRef.take then
+    -- someone else was quicker!
+    eligibleHeaderDeclsRef.set eligibleHeaderDecls
+    return eligibleHeaderDecls
+  let (_, eligibleHeaderDecls) :=
+    StateT.run (m := Id) (s := {}) do
+      -- `map₁` are the header decls
+      env.constants.map₁.forM fun declName c => do
         modify fun eligibleHeaderDecls =>
-          if allowCompletion headerEnv declName then
+          if allowCompletion env declName then
             eligibleHeaderDecls.insert declName c
           else
             eligibleHeaderDecls
+  eligibleHeaderDeclsRef.set (some eligibleHeaderDecls)
+  return eligibleHeaderDecls
 
 /-- Iterate over all declarations that are allowed in completion results. -/
 private def forEligibleDeclsM [Monad m] [MonadEnv m] [MonadLiftT (ST IO.RealWorld) m]
-    (f : Name → ConstantInfo → m PUnit) : m PUnit := do
+    [MonadLiftT IO m] (f : Name → ConstantInfo → m PUnit) : m PUnit := do
   let env ← getEnv
-  (← eligibleHeaderDeclsRef.get).forM f
+  (← getEligibleHeaderDecls (← getEnv)).forM f
   -- map₂ are exactly the local decls
   env.constants.map₂.forM fun name c => do
     if allowCompletion env name then
       f name c
 
 /-- Checks whether this declaration can appear in completion results. -/
-private def allowCompletion [Monad m] [MonadEnv m] [MonadLiftT (ST IO.RealWorld) m]
-    (declName : Name) : m Bool := do
-  let env ← getEnv
-  if (← eligibleHeaderDeclsRef.get).contains declName then
-    return true
-  if env.constants.map₂.contains declName && Lean.Meta.allowCompletion env declName then
-    return true
-  return false
+private def allowCompletion (env : Environment) (eligibleHeaderDecls : EligibleHeaderDecls)
```

Comment:
My general rule of thumb for parameter order is "parameters that are unchanged for longer go first". By that rule, `eligibleHeaderDecls` should go before `env`.